libr/anal/esil.c

```diff
@@ -544,6 +544,48 @@ R_API int r_anal_esil_reg_read(RAnalEsil *esil, const char *regname, ut64 *num,
 	return ret;
 }
 
+// sign extension operator for use in idiv, imul, movsx* 
+// and other instructions involving signed values, extends n bit value to 64 bit value
+// example : >"ae 8,0x81,~" ( <src bit width>,<value>,~ )
+// output  : 0xffffffffffffff81
+static bool esil_signext(RAnalEsil *esil) {
+	ut64 src, dst;
+
+	char *p_src = r_anal_esil_pop (esil);
+	if (!p_src) {
+		return false;
+	}
+
+	if (!r_anal_esil_get_parm (esil, p_src, &src)) {
+		ERR ("esil_of: empty stack");
+		free (p_src);
+		return false;
+	} else {
+		free (p_src);
+	}
+
+	char *p_dst = r_anal_esil_pop (esil);
+	if (!p_dst) {
+		return false;
+	}
+
+	if (!r_anal_esil_get_parm (esil, p_dst, &dst)) {
+		ERR ("esil_of: empty stack");
+		free (p_dst);
+		return false;
+	} else {
+		free (p_dst);
+	}
+
+	ut64 m = 0;
+	if (dst < 64) {
+		m = 1ULL << (dst - 1);
+	}
+
+	// dst = (dst & ((1U << src_bit) - 1)); // clear upper bits
+	return r_anal_esil_pushnum (esil, ((src ^ m) - m));
```

Comment:
so technically this is just xor and subtraction?

Reply:
Yea it can be done with existing operators and has been done in a bunch of ad hoc ways in different archs. But that blows up the size and complexity of expressions and I think there is a lot of value in just having the operator so we can standardize it.